server/dvo_handlers.go

```diff
@@ -0,0 +1,298 @@
+				Recommendations: int(workload.Recommendations),
+				Objects:         int(workload.Objects),
+				ReportedAt:      string(workload.ReportedAt),
+				LastCheckedAt:   string(workload.LastCheckedAt),
+			},
+			// TODO: fill RecommendationsHitCount map efficiently instead of processing the report again every time
+		})
+	}
+
+	return
+}
+
+// getWorkloadsForNamespace retrieves data about a single namespace within a cluster
+func (server *HTTPServer) getWorkloadsForNamespace(writer http.ResponseWriter, request *http.Request) {
+	tStart := time.Now()
+
+	orgID, ok := readOrgID(writer, request)
+	if !ok {
+		// everything has been handled
+		return
+	}
+
+	clusterName, successful := readClusterName(writer, request)
+	if !successful {
+		// everything has been handled already
+		return
+	}
+
+	namespaceID, err := readNamespace(writer, request)
+	if err != nil {
+		return
+	}
+
+	log.Debug().Int(orgIDStr, int(orgID)).Str("namespaceID", namespaceID).Msgf("getWorkloadsForNamespace cluster %v", clusterName)
+
+	workload, err := server.StorageDvo.ReadWorkloadsForClusterAndNamespace(orgID, clusterName, namespaceID)
+	if err != nil {
+		log.Error().Err(err).Msg("Errors retrieving DVO workload recommendations from storage")
+		handleServerError(writer, err)
+		return
+	}
+
+	processedWorkload := server.processSingleDVONamespace(workload)
+
+	log.Info().Uint32(orgIDStr, uint32(orgID)).Msgf(
+		"getWorkloadsForNamespace took %s", time.Since(tStart),
+	)
+	err = responses.SendOK(writer, responses.BuildOkResponseWithData("workloads", processedWorkload))
+	if err != nil {
+		log.Error().Err(err).Msg(responseDataError)
+	}
+}
+
+// processSingleDVONamespace processes a report, filters out mismatching namespaces, returns processed results
+func (server *HTTPServer) processSingleDVONamespace(workload types.DVOReport) (
+	processedWorkloads WorkloadsForCluster,
+) {
+	processedWorkloads = WorkloadsForCluster{
+		Cluster: Cluster{
+			UUID: workload.ClusterID,
+		},
+		Namespace: Namespace{
+			UUID: workload.NamespaceID,
+			Name: workload.NamespaceName,
+		},
+		Metadata: Metadata{
+			Recommendations: int(workload.Recommendations),
+			Objects:         int(workload.Objects),
+			ReportedAt:      string(workload.ReportedAt),
+			LastCheckedAt:   string(workload.LastCheckedAt),
+		},
+		Recommendations: []DVORecommendation{},
+	}
+
+	var dvoReport types.DVOMetrics
+	// remove doubled escape characters due to improper encoding during storage
+	s := strings.Replace(workload.Report, "\\", "", -1)
+
+	err := json.Unmarshal(json.RawMessage(s), &dvoReport)
+	if err != nil {
+		log.Error().Err(err).Msg("error unmarshalling full report")
+		return
+	}
+
+	for _, recommendation := range dvoReport.WorkloadRecommendations {
+		filteredObjects := make([]DVOObject, 0)
+		for i := range recommendation.Workloads {
+			object := &recommendation.Workloads[i]
+
+			// filter out other namespaces
+			if object.NamespaceUID != processedWorkloads.Namespace.UUID {
```

Comment:
probably not needed, since we're already doing this filtering in the SQL query. Or am I misunderstanding something?

Reply:
yeah, this is filtering the objects (workloads) themselves inside each recommendation. Because the whole report is not grouped by namespace.. the whole report always contains data for all namespaces that we have to filter out since we only want 1 namespace. This is something we could refactor to only store the relevant part of the report but for now (until GA) it's not a bad idea to be storing the complete report